### ISSUE TITLE
Problem with the Atomium instalation on windows?

### ISSUE THREAD

**Issue by chromascope** [BradyAJohnston/MolecularNodes#4]
I get the error below when trying to import a pdb-file, online or localy.

I have followed the steps in the instalation guide:
I run Blender as an administrator on Windows 10, and after installing the add-on, i click the button to install Atomium. I get a confirmation that Atomium is successfully installed.

In the MolecularNodes panel i click the "Download" / "Open" button and get the error below. I assume the it is because the Atomium library is not properly installed.

```
Python: Traceback (most recent call last):
  File "C:\Users\Jonas\AppData\Roaming\Blender Foundation\Blender\3.1\scripts\addons\molecular_nodes\__init__.py", line 1401, in execute
    sna_func_import_structure_EE5B6(bpy.context.scene.sna_pdb_code,
  File "C:\Users\Jonas\AppData\Roaming\Blender Foundation\Blender\3.1\scripts\addons\molecular_nodes\__init__.py", line 1134, in sna_func_import_structure_EE5B6
    pdb = atomium.fetch(pdb_id)
UnboundLocalError: local variable 'atomium' referenced before assignment

location: <unknown location>:-1
```

I am very curious to try this amazing add-on, hope that you can help resolve the issue.

**Comment by BradyAJohnston** [comment]
Hmmm, unsure why it might be failing to install. 

Can you try opening blender, going to the scripting workspace (top of the window), creating a new script, paste this code in and press run? 
```python
import subprocess
import sys
import os
 
# path to python.exe
python_exe = os.path.realpath(sys.executable)
 
# upgrade pip
subprocess.call([python_exe, "-m", "ensurepip"])
subprocess.call([python_exe, "-m", "pip", "install", "--upgrade", "pip"])
 
# install required packages
subprocess.call([python_exe, "-m", "pip", "install", "atomium"])
```

![image](https://user-images.githubusercontent.com/36021261/165481443-e5c161e6-f492-47c0-aede-efe3ffcc3ae3.png)


If you get errors, open the system console and paste in the errors that you are getting. 

![image](https://user-images.githubusercontent.com/36021261/165481668-a27b79dd-07fa-4322-968e-0f01a92f651d.png)

**Comment by chromascope** [comment]
Hi, thanks for quick response. :-) I get the following feedback from the system console when running the script as you suggest. I still get the same error when trying to import a file in the panel.
```
Looking in links: c:\Users\Jonas\AppData\Local\Temp\tmpimlxdg9m
Requirement already satisfied: setuptools in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (58.1.0)
Requirement already satisfied: pip in c:\users\jonas\appdata\roaming\python\python310\site-packages (22.0.4)
Requirement already satisfied: pip in c:\users\jonas\appdata\roaming\python\python310\site-packages (22.0.4)
Requirement already satisfied: atomium in c:\users\jonas\appdata\roaming\python\python310\site-packages (1.0.11)
Requirement already satisfied: paramiko in c:\users\jonas\appdata\roaming\python\python310\site-packages (from atomium) (2.10.4)
Requirement already satisfied: numpy in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from atomium) (1.22.0)
Requirement already satisfied: requests in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from atomium) (2.27.1)
Requirement already satisfied: valerius in c:\users\jonas\appdata\roaming\python\python310\site-packages (from atomium) (0.2.0)
Requirement already satisfied: rmsd in c:\users\jonas\appdata\roaming\python\python310\site-packages (from atomium) (1.4)
Requirement already satisfied: msgpack in c:\users\jonas\appdata\roaming\python\python310\site-packages (from atomium) (1.0.3)
Requirement already satisfied: bcrypt>=3.1.3 in c:\users\jonas\appdata\roaming\python\python310\site-packages (from paramiko->atomium) (3.2.0)
Requirement already satisfied: cryptography>=2.5 in c:\users\jonas\appdata\roaming\python\python310\site-packages (from paramiko->atomium) (37.0.0)
Requirement already satisfied: pynacl>=1.0.1 in c:\users\jonas\appdata\roaming\python\python310\site-packages (from paramiko->atomium) (1.5.0)
Requirement already satisfied: six in c:\users\jonas\appdata\roaming\python\python310\site-packages (from paramiko->atomium) (1.16.0)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from requests->atomium) (1.26.8)
Requirement already satisfied: certifi>=2017.4.17 in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from requests->atomium) (2021.10.8)
Requirement already satisfied: charset_normalizer~=2.0.0 in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from requests->atomium) (2.0.10)
Requirement already satisfied: idna<4,>=2.5 in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from requests->atomium) (3.3)
Requirement already satisfied: scipy in c:\users\jonas\appdata\roaming\python\python310\site-packages (from rmsd->atomium) (1.8.0)
Requirement already satisfied: cffi>=1.1 in c:\users\jonas\appdata\roaming\python\python310\site-packages (from bcrypt>=3.1.3->paramiko->atomium) (1.15.0)
Requirement already satisfied: pycparser in c:\users\jonas\appdata\roaming\python\python310\site-packages (from cffi>=1.1->bcrypt>=3.1.3->paramiko->atomium) (2.21)21)
```

**Comment by BradyAJohnston** [comment]
What happens if you have an empty script that just contains the following:

```python
import atomium
atomium.fetch("1bna")
```

**Comment by chromascope** [comment]
I get this in the console:
```
Traceback (most recent call last):
  File "\Text", line 1, in <module>
ModuleNotFoundError: No module named 'atomium'
Error: Python script failed, check the message in the system console
```

**Comment by BradyAJohnston** [comment]
Hmm it seems that when installing, the package is installed to a system python library, rather than the blender python library. The requirements are satisfied. Could maybe try reinstalling Blender and see if that fixes the problem. There are something python path problems that seem to be going on.

**Comment by BradyAJohnston** [comment]
Can you both run this and see what it says? 
```python
import subprocess
import sys
import os
 
python_exe = os.path.realpath(sys.executable)
subprocess.call([python_exe, "-m", "pip", "show", "atomium"])
```

**Comment by emalacs** [comment]
Now I have this error:

`Read prefs: C:\Users\odino\AppData\Roaming\Blender Foundation\Blender\3.1\config\userpref.blend
Name: atomium
Version: 1.0.11
Summary: A molecular modeller and file parser.
Home-page: https://atomium.samireland.com
Author: Sam Ireland
Author-email: mail@samireland.com
License: MIT
Location: c:\users\odino\appdata\roaming\python\python310\site-packages
Requires: msgpack, numpy, paramiko, requests, rmsd, valerius
Required-by:
`

**Comment by emalacs** [comment]
I have this error now:

```
Read prefs: C:\Users\odino\AppData\Roaming\Blender Foundation\Blender\3.1\config\userpref.blend
Name: atomium
Version: 1.0.11
Summary: A molecular modeller and file parser.
Home-page: https://atomium.samireland.com
Author: Sam Ireland
Author-email: mail@samireland.com
License: MIT
Location: c:\users\odino\appdata\roaming\python\python310\site-packages
Requires: msgpack, numpy, paramiko, requests, rmsd, valerius
Required-by:
```

**Comment by chromascope** [comment]
Same here:
```
Name: atomium
Version: 1.0.11
Summary: A molecular modeller and file parser.
Home-page: https://atomium.samireland.com
Author: Sam Ireland
Author-email: mail@samireland.com
License: MIT
Location: c:\users\jonas\appdata\roaming\python\python310\site-packages
Requires: msgpack, numpy, paramiko, requests, rmsd, valerius
Required-by:
```

**Comment by BradyAJohnston** [comment]
Okay I think I have a way that might fix it. I'm not sure why the error is popping up in the first place, but it seems that `os.path.realpath(sys.executable)` is picking up your system installation of python and not the blender installation of python. 

If you can navigate to your installation of Blender, and find the `python.exe`: 
```bash
*PATHTOBLENDER*\Blender\3.1\python\bin\python.exe
```

You will use that inside of the script, but making sure that you replace all single backslashes with double backslashes. Such as below : 
```python
python_exe = "*PATHTOBLENDER*\\Blender\\3.1\\python\\bin\\python.exe"
```

And then use it in the code that installs atomium: 
```python
import subprocess
import sys
import os
 
python_exe = "*PATHTOBLENDER*\\Blender\\3.1\\python\\bin\\python.exe"

subprocess.call([python_exe, "-m", "ensurepip"])
subprocess.call([python_exe, "-m", "pip", "install", "--upgrade", "pip"])
 
subprocess.call([python_exe, "-m", "pip", "install", "atomium"])
```

Running that inside of the console or as a script in Blender, should hopefully install `atomium` in a library that Blender is able to access, which should enable you to run the download script.

**Comment by chromascope** [comment]
Hmm, looks like it is still installing in the default location unfortunately.

```
Looking in links: c:\Users\Jonas\AppData\Local\Temp\tmpnnhyk_e9
Requirement already satisfied: setuptools in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (58.1.0)
Requirement already satisfied: pip in c:\users\jonas\appdata\roaming\python\python310\site-packages (22.0.4)
Requirement already satisfied: pip in c:\users\jonas\appdata\roaming\python\python310\site-packages (22.0.4)
Requirement already satisfied: atomium in c:\users\jonas\appdata\roaming\python\python310\site-packages (1.0.11)
Requirement already satisfied: numpy in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from atomium) (1.22.0)
Requirement already satisfied: rmsd in c:\users\jonas\appdata\roaming\python\python310\site-packages (from atomium) (1.4)
Requirement already satisfied: msgpack in c:\users\jonas\appdata\roaming\python\python310\site-packages (from atomium) (1.0.3)
Requirement already satisfied: valerius in c:\users\jonas\appdata\roaming\python\python310\site-packages (from atomium) (0.2.0)
Requirement already satisfied: requests in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from atomium) (2.27.1)
Requirement already satisfied: paramiko in c:\users\jonas\appdata\roaming\python\python310\site-packages (from atomium) (2.10.4)
Requirement already satisfied: six in c:\users\jonas\appdata\roaming\python\python310\site-packages (from paramiko->atomium) (1.16.0)
Requirement already satisfied: pynacl>=1.0.1 in c:\users\jonas\appdata\roaming\python\python310\site-packages (from paramiko->atomium) (1.5.0)
Requirement already satisfied: cryptography>=2.5 in c:\users\jonas\appdata\roaming\python\python310\site-packages (from paramiko->atomium) (37.0.0)
Requirement already satisfied: bcrypt>=3.1.3 in c:\users\jonas\appdata\roaming\python\python310\site-packages (from paramiko->atomium) (3.2.0)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from requests->atomium) (1.26.8)
Requirement already satisfied: certifi>=2017.4.17 in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from requests->atomium) (2021.10.8)
Requirement already satisfied: charset_normalizer~=2.0.0 in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from requests->atomium) (2.0.10)
Requirement already satisfied: idna<4,>=2.5 in c:\program files\blender foundation\blender 3.1\3.1\python\lib\site-packages (from requests->atomium) (3.3)
Requirement already satisfied: scipy in c:\users\jonas\appdata\roaming\python\python310\site-packages (from rmsd->atomium) (1.8.0)
Requirement already satisfied: cffi>=1.1 in c:\users\jonas\appdata\roaming\python\python310\site-packages (from bcrypt>=3.1.3->paramiko->atomium) (1.15.0)
Requirement already satisfied: pycparser in c:\users\jonas\appdata\roaming\python\python310\site-packages (from cffi>=1.1->bcrypt>=3.1.3->paramiko->atomium) (2.21)
```

**Comment by BradyAJohnston** [comment]
Hmm okay it so it's definitely a `$PYTHONPATH` issue. I honestly have no idea how to solve it currently. It seems that Blender's `pip` is aware of your system package libraries, and considers it successfully installed, but the Blender python seems to be unable to find those packages. It will likely take someone who knows their way around python better to figure this out. I'll do some reading and see what I can find but I am stumped currently.

**Comment by emalacs** [comment]
Same of @chromascope. Actually I was trying on linux and I still have the same error during the pdb opening.

**Comment by emalacs** [comment]
Ok, so I just tried to use Blender 3.0.1 and it works!

![image](https://user-images.githubusercontent.com/57134605/165512255-4af7b9c4-bbab-448e-928b-9fdc4658a366.png)

**Comment by BradyAJohnston** [comment]
Okay that is an interesting development! 

Some of the nodes might not work in `3.0.1`, but once you have imported the structure you should be able to open it up inside of `3.1` without issue. Atomium is only used for the initial import step.

**Comment by giovanna-deluca** [comment]
Hi! I am experiencing the same problems and getting the same outputs from the above scripts.
I am using Blender 3.1.2 under Arch Linux, and I am a newbie both with Blender and the OS...

Still trying to locate python hehehe


> 
> If you can navigate to your installation of Blender, and find the `python.exe`:
> 
> *PATHTOBLENDER*\Blender\3.1\python\bin\python.exe

**Comment by BradyAJohnston** [comment]
@giovanna-deluca try downloading and installing Blender `3.0` from here: https://download.blender.org/release/Blender3.0/ and then import the protein with the panel, and then save the file and open inside of `3.1` as a workaround. Might be that it's a problem with Blender 3.1.2

**Comment by BradyAJohnston** [comment]
@giovanna-deluca thanks for the update! (and thanks to your husband) I am glad it is working properly, it seems that on Linux you require elevate permissions to install it as well. I'm not a linux desktop user - is there a way to launch programs 'As Administrator' like on Windows? Or will people have to just use `sudo` via a terminal to install the package?

**Comment by fpqhms** [comment]
I was able to install everything including addon and Atomium on my mac, but have trouble to fetch any pdb file either from the PDB database or my local folder. The error messages are as following:
 
_Python: Traceback (most recent call last):
  File "/Users/peiqiangfeng/Library/Application Support/Blender/3.1/scripts/addons/molecular_nodes/__init__.py", line 1401, in execute
    sna_func_import_structure_EE5B6(bpy.context.scene.sna_pdb_code,
  File "/Users/peiqiangfeng/Library/Application Support/Blender/3.1/scripts/addons/molecular_nodes/__init__.py", line 1134, in sna_func_import_structure_EE5B6
    pdb = atomium.fetch(pdb_id)
UnboundLocalError: local variable 'atomium' referenced before assignment
 
location: <unknown location>:-1_

I am using macOS and my Blender is 3.1.2. Looks like it is related with the Blender version because I can do it in Blender 3.0.1.

**Comment by BradyAJohnston** [comment]
Hi @fpqhms thanks for reporting that you are getting this issue too. Interesting that it is happening on both Windows and Mac. It seems that there is something strange about the `3.1.2` release of Blender that is causing problems with installing `atomium`. The workaround currently seems downloading structures through the earlier version. Might be unfixable until the next blender release if this is a blender-centric issue.

**Comment by BradyAJohnston** [comment]
From the [Release Notes](https://wiki.blender.org/wiki/Reference/Release_Notes/3.1/Python_API): 

> ## Python has been upgraded from version 3.9 to 3.10, there are some changes that may impact script authors. 

Might be that there is a change in how python handles these things itself, that is causing the issue.

**Comment by chromascope** [comment]
> Hi @fpqhms thanks for reporting that you are getting this issue too. Interesting that it is happening on both Windows and Mac. It seems that there is something strange about the `3.1.2` release of Blender that is causing problems with installing `atomium`. The workaround currently seems downloading structures through the earlier version. Might be unfixable until the next blender release if this is a blender-centric issue.

I don't find any issues with the add-on on MacOS. I tried both Blender 3.1 and 3.1.2, and both install atomium without any issues for me.

**Comment by BradyAJohnston** [comment]
@chromascope thank you also for reporting that it is working! It seems to be inconsistent in the error, which is frustrating because I can't seem to replicate it myself to try and troubleshoot. I am working on a more robust method of installing `atomium` that may help, we shall. It's good to see that it's working okay for some though!

**Comment by BradyAJohnston** [comment]
Thanks @giovanna-deluca for that post! Could you try downloading the 0.3.7 release, and seeing if it fixes it? https://github.com/BradyAJohnston/MolecularNodes/releases/tag/atomium-fix

You will likely want to uninstall Blender, and uninstall atomium with pip, then re-install blender, reinstall the addon and try the "install atomium" button again to see if it works

**Comment by giovanna-deluca** [comment]
@BradyAJohnston thanks for the update! The issue is resolved in my case.

As per your request, I uninstalled Blender and atomium (both as user and as root).
Reinstalled Blender 3.1.2. Added your release 0.3.7 and installed atomium from the related button.
I confirmed atomium is now present only in the user site-packages path by listing the python modules from bash.
Finally, I successfully imported the pdb file via your add-on!

**Comment by BradyAJohnston** [comment]
This is _excellent_ news! Hopefully other people have similar kinds of success, we may have a solution!

**Comment by emalacs** [comment]
I tried to reinstall Blender removing Atomium and try the 0.3.7 release. It kinda works meaning that I am able to install atomium and import a pdb. However, as soon as I close Blender and reopen it, I need to reinstall atomium again for some weird reason. But it is not necessary to open Blender as administrator this time.

**Comment by giovanna-deluca** [comment]
@BradyAJohnston, I can confirm @emalacs observation.
On running Blender again and trying to import the pdb file, I received:

`Python: Traceback (most recent call last):
  File "/home/giov/.config/blender/3.1/scripts/addons/molecular_nodes/__init__.py", line 684, in execute
    sna_func_import_structure_EE5B6(
  File "/home/giov/.config/blender/3.1/scripts/addons/molecular_nodes/__init__.py", line 1213, in sna_func_import_structure_EE5B6
    pdb = atomium.fetch(pdb_id)
UnboundLocalError: local variable 'atomium' referenced before assignment

location: <unknown location>:-1

`

**Comment by giovanna-deluca** [comment]
@BradyAJohnston we tested something, and it appears to have worked:

We modified your file __init__.py so that the inclusion of the user site-packages path occurs in the register() function.
We added the following code:

```
def verify_user_sitepackages():
    import site
    import sys
    usersitepackagespath = site.getusersitepackages()
    if os.path.exists(usersitepackagespath) and usersitepackagespath not in sys.path:
        sys.path.append(usersitepackagespath)
```
at the beginning of the file and then invoked the above inside the register() function.

Then, I disabled/enabled the add-on and it worked.
Since then, on closing and reopening Blender your add-on works.

Hope it helps... we are not python experts...

**Comment by BradyAJohnston** [comment]
Ah yes I the path will just need re-registering every time I try to use it. 

The [0.3.8 release](https://github.com/BradyAJohnston/MolecularNodes/releases/tag/atomium-fix-2) should fix that! Thanks for all of the detailed feedback so far :)

**Comment by BradyAJohnston** [comment]
This is excellent news! I will leave this issue open for now, and will close with the `0.4` release.

**Comment by BradyAJohnston** [comment]
fixed in [0.4 release](https://github.com/BradyAJohnston/MolecularNodes/releases/tag/0.4)
